### PR TITLE
New Poisson RNG

### PR DESCRIPTION
**Pull request by jkreschke**
Implemented a new Poisson RNG not using the direct method, so runs efficiently for all lambda.

### CODE REVIEW

**Comment by jkreschke** [comment]
Will @rls347, @mchinita 
please review this PR, thanks!

**Comment by mchinita** [comment]
Hi Everyone, 

I think there is a typo in L1063: alphinv = 1.1239_r8 + **1.328**_r8/(b - 3.4_r8)

I think the number is 1.1328. It's just missing "1" before "328". Do you agree?

**Comment by jkreschke** [comment]
Hi Maria,

Yes, you are correct. Thanks for double checking the paper!

Jake

**Comment by mchinita** [comment]
Hi Jake,

Sure! 

One genuine question: our lambdas are usually smaller than 10 (e.g. delta_Z/L0 = 150/100 = 1.5). Do you think this algorithm works as well as it does for lambdas > 10? If I understand correctly, the values in Table 1 were optimized for lambdas >= 10. 

Thanks for coding the new algorithm! I should probably integrate it in SHOC_MF as well, but I will let you know before. 

Maria

**Review by mikaelwitte — CHANGES_REQUESTED** [review]
In the preamble, state which algorithm you use from the Hormann reference (PTRS, right?)

Line 1063: Agreed with Maria, second constant in alphinv should be 1.1328_r8. 

Line 1070: The final constant here is given as 0.445 in the Hormann paper. Just checking whether this was intentional.

Everything else looks good, thanks Jake!

**Comment by jkreschke** [comment]
Hi everyone,

Thanks for looking for over the code and sorry for the confusion, I should have specified that the PTRS algorithm is being used. 

I didn’t actually try testing the algorithm for lambda < 10 to see what happens. The Knuth algorithm is inexpensive for small lambda so we just use it in that case.

The k value differs between the two algorithms presented in Hormann (for reasons unknown to me). In the first algorithm the constant is 0.445 but in the PTRS version it is 0.43.

Best,
Jake

**Comment by mchinita** [comment]
Indeed Jake. I forgot for a second that smaller values of L0 and thus bigger lambdas were the reason for this new algorithm. If I remember correctly (Adam correct me if I'm wrong), Sc are better captured using very small L0 with the old entrainment formulation. But with the new TKE enhanced entrainment formulation, we are able to properly represent Sc while keep using the L0 = a*sqrt(ztop) which produces larger L0 values, and thus smaller lambdas. So yes, in that case, we can still use the old algorithm.

Thanks!
Maria

File: src/physics/cam/clubb_mf.F90 [hunk 1]
```diff
@@ -992,12 +986,31 @@ subroutine poisson(nz,nup,lambda,poi,state)
...
+  ! Interface for the two poisson rng subroutines
+  ! chooses the appropriate subroutine based on the value of lambda
+  !**********************************************************************
+   use shr_RandNum_mod, only: ShrKissRandGen
+
+       type(ShrKissRandGen), intent(inout) :: kiss_gen
+       real(r8),             intent(in)    :: lambda
+       integer,              intent(out)   :: kout
+
+       if (lambda < 10) then
```
Comment (adamrher) [inline src/physics/cam/clubb_mf.F90]: Jake can you change this to 10._r8 to make it compiler proof?

File: src/physics/cam/clubb_mf.F90 [hunk 2]
```diff
@@ -1027,6 +1040,59 @@ subroutine knuth(kiss_gen,lambda,kout)
 
   end subroutine knuth
 
+  subroutine hormann(kiss_gen,lambda,kout)
+  !**********************************************************************
+  ! Discrete random poisson
+  ! Implements PTRS algorithm from W. Hormann
```
Comment (adamrher) [inline src/physics/cam/clubb_mf.F90]: What does PTRS stand for? If it's not too long, I'd like to see the unabbreviated version here.

File: src/physics/cam/clubb_mf.F90 [hunk 3]
```diff
@@ -1027,6 +1040,59 @@ subroutine knuth(kiss_gen,lambda,kout)
...
+         call kiss_gen%random(U)
+         call kiss_gen%random(V)
+         U(1,1) = U(1,1) - 0.5_r8
+         us = 0.5_r8 - abs(U(1,1))
+         k = floor( (2._r8*a/us + b)*U(1,1) + lambda + 0.43_r8 )
+         if (us >= 0.07_r8 .and.  V(1,1) <= vr) then
+            kout = k
+            exit
+         end if
+         if (k <= 0 .or. (us < 0.013_r8 .and. V(1,1) > us)) then
```
Comment (adamrher) [inline src/physics/cam/clubb_mf.F90:1076]: similarly change this to k <= 0._r8

File: src/physics/cam/clubb_mf.F90 [hunk 4]
```diff
@@ -1027,6 +1040,59 @@ subroutine knuth(kiss_gen,lambda,kout)
...
+            cycle
+         end if
+         ! compute log(k!). If k >=10 use stirling's approximation
+         if (k < 10) then
```
Comment (adamrher) [inline src/physics/cam/clubb_mf.F90:1080]: same here

**Review by adamrher — APPROVED** [review]
Looks good. I will approve assuming you address my comments.

File: src/physics/cam/clubb_mf.F90 (hunk 4 above)
Comment (adamrher) [reply src/physics/cam/clubb_mf.F90:1080]: I also take this back! k is an integer.

File: src/physics/cam/clubb_mf.F90 (hunk 3 above)
Comment (adamrher) [reply src/physics/cam/clubb_mf.F90:1076]: oh wait, I take this back. k is an integer.

**Comment by jkreschke** [comment]
Done!

**Pull request merged**